Add support to test multi service API version (#12719)
Azure/azure-sdk-for-js: sdk/test-utils/multi-version/README.md

## Azure SDK for JavaScript's test-utils-multi-version

The Azure SDK for JavaScript is composed of a multitude of repositories that attempt to deliver a

> common, homogenous SDK to make use of all of the services that Azure can provide. One of the

Reviewer: ```suggestion
common, homogenous SDK to make use of all of the services that Azure offers. One of the
```
?

Author: This was copied from test-utils-recorder...

Reviewer: Ouch!

Author: Won't fix, to be consistent.

> promises the Azure SDK libraries is to support the last N service API versions.

Reviewer: ```suggestion
promises of the Azure SDK libraries is to support the last N service API versions.
```

Author: fixed

This non-shipping library `@azure/test-utils-multi-version` attempts to add testing support for
libraries that supports multiple service API versions. It is supposed to be added only as a
devDependency and should be used only for the tests of an SDK library.

The idea employed in this library is inspired by [mocha-tags](https://www.npmjs.com/package/mocha-tags).

## Key concepts

- Our guideline recommends that service client supporting multiple service API version takes an API
  version via the `serviceVersion` property of its constructor options bag `*ClientOptions`.
- Service API versions supported by the SDK library under test is specified for a top-level Mocha
  `describe()` test suite. We loop through them and run the test suite on each version.
  - Testing of multi-service-version support is only done in `live` test mode. For other test modes
    (`record`, `soft-record`, and `playback`), tests are only running for the latest version, or a
    chosen version specified via options in case the latest version is not the latest stable version.
- Optionally a list or range of versions can be specified for a nested `describe()` test suite or

>   `it()` test case. If the version being tested is not in the range of versions supported by this
>   test suite/case then the test suite/case is skipped.

Reviewer: ```suggestion
  `it()` test case. If the version being tested is not in the range of versions supported by this test suite/case, then the test suite/case is skipped.
```

Author: fixed

git clone https://github.com/Azure/azure-sdk-for-js/
```

Having cloned this repository, let's set it up by running the following rush commands:

```bash
cd azure-sdk-for-js
rush update
rush install
rush build
```

This will optimistically assume you're in a fresh clone.

From this point forward, we'll assume that you're developing (perhaps contributing!) to one of the
`azure-sdk-for-js`'s libraries. So, your next step is to change directory to the path relevant to
your project. Let's say you want to add the `@azure/test-utils-multi-version` package to
`@azure/keyvault-keys`, you'll be doing the following:

```bash
cd sdk/keyvault/keyvault-keys
```

Once there, you can add the `test-utils-multi-version` package by changing your package.json
to include the following line in the `devDependencies` section:

```bash
{
  // ... your package.json properties
  "devDependencies": {
    // ... your devDependencies
    "@azure/test-utils-multi-version": "^1.0.0",
    // ... more of your devDependencies
  },
  // ... more of your package.json properties
}
```

After that, we recommend you to update rush and install the dependencies again, as follows:

```bash

> rush update && rush install

Reviewer: `rush install` may not be needed?

Author: Sure, again copied from test-utils-recorder.

Author: removed

```

And you're ready to test your library for multiple service API versions!

## Examples

### Import functions

```javascript
import { supports, versionsToTest, SupportedVersions } from "@azure/test-utils-multi-version";
```

### Wrap top-level test suite

Wrap a top level `describe()` of a test file to enable testing for multiple versions for that test suite:

```javascript
const serviceApiVersions = ["7.0", "7.1"] as const;
versionsToTest(["7.0", "7.1"]).forEach((serviceVersion) => {
  describe("Keys client - list keys in various ways", async function() {
   // ...
  }
}
```

`versionsToTest()` takes in a list of versions that are supported by the library. it checks the

> current `TEST_MODE` then either runs tests for all supported versions in `live` mode, or runs tests
> for just one version (default version if specified in options, or latest version in the list). The

Reviewer: ```suggestion
current `TEST_MODE` then either runs tests for all supported versions in `live` mode, or runs tests
for just one version (default version if specified in options, or latest version in the list) in the other test modes. The
```
?

Author: fixed

test suite, and its nested test suites/test cases all have access to the current version being
tested - `serviceVersion` - so they can verify different behavior/expectation if any for different
service API versions.

> The code to construct clients for testing also needs update to pass the `serviceVersion` to client
> constructors.

Reviewer: ```suggestion
The code to construct clients for testing also needs an update to pass the `serviceVersion` to the client
constructors.
```

Author: added "an"
